src/app.tsx

```diff
@@ -47,6 +47,14 @@ export interface FolderFileInfo extends FileInfo {
     to: string | null,
 }
 
+interface FilesContextProvider {
+    addAlert?: (title: string, variant: AlertVariant, key: string) => void,
```

Comment:
For my education: That means that `addAlert` is a property which can be set to a function or be `undefined`, right? I. e. "optional property'. I suppose this is so that you can initialize it below without specifying an actual callback, in `const initialContextState`. This is ugly, can this be avoided, so that this is reliably existing? Perhaps with a default implementation that uses `console.warn()`?

If that's the standard pattern, then nevermind. But it is only set once in app.tsx (I assume that's what `<FilesContext.Provider value={{ addAlert }}>` does?) Can't the `FilesContext` initialization be moved into `Application` somehow so that it can be initialized once and for all?

Reply:
> For my education: That means that `addAlert` is a property which can be set to a function or be `undefined`, right? I. e. "optional property'. I suppose this is so that you can initialize it below without specifying an actual callback, in `const initialContextState`. This is ugly, can this be avoided, so that this is reliably existing? Perhaps with a default implementation that uses `console.warn()`?

Yes this is ugly, so the `?` is there because initially it is empty. We can drop it and provide a default `console.warn`. BTW, this is an issue in Cockpit as well but we have no Typescript there.

There we just call `createContext` which actually requires an argument but ends up getting undefined as default state. So I've re-adjusted it locally and this makes it a bit nicer as it drops the `interface`

```
+export const FilesContext = React.createContext({ addAlert: () => console.warn("FilesContext not initialized") });
```

> If that's the standard pattern, then nevermind. But it is only set once in app.tsx (I assume that's what `<FilesContext.Provider value={{ addAlert }}>` does?) Can't the `FilesContext` initialization be moved into `Application` somehow so that it can be initialized once and for all?

I don't believe you can call `createContext` inside of `Application` as it will re-render.

Reply:
Thanks for the explanation! The console.warn() seems fine to me then.